2022/Q3/artifacts/PoS-tla/Staking.tla

```diff
@@ -113,104 +143,257 @@ Delegate(sender, validator, amount) ==
-                     IF n >= PipelineLength /\ user = sender
-                     THEN bonded[n, user] + amount
-                     ELSE bonded[n, user]
-                    ]
-          /\ UNCHANGED unbonded
-
-
-\* unbond `amount` tokens from a validator
+          /\ posAccount' = posAccount + amount
+          /\ bonded' = [ bonded EXCEPT ![sender, validator] = @ \union {[ id |-> idBonds,
+                                                                          epoch |-> epoch + PipelineLength,
+                                                                          amount |-> amount ]}]
+          /\ idBonds' = idBonds + 1
+          \* updates totalDeltas from PipelineLength to UnbondingLength 
+          /\ totalDeltas' = [ n \in 0..2*UnbondingLength, val \in ValidatorAddrs |-> 
+                              IF n >= UnbondingLength + PipelineLength /\ val = validator
+                              THEN totalDeltas[n, val] + amount
+                              ELSE totalDeltas[n, val]
+                            ]
+
+\* Follows the specification: it counts all bonds, even those that have not been materialized yet.
+\* PAY ATTENTION TO IT
+\* @type: (ADDR, ADDR) => Int;
+TotalBonds(sender, validator) == LET 
+                                 \* @type: (Int, BOND) => Int;
+                                 F(total, bond) == total + bond.amount
+                                 IN ApaFoldSet(F, 0, bonded[sender, validator])
+
+\* @type: (Set(BOND), Int, Int) => [ remaining: Int, set: Set(UNBOND), id: Int ];
+ComputedUnbonds(setBonds, amount, e) == LET 
+                                         \* @type: ([ remaining: Int, set: Set(UNBOND), id: Int ], BOND) => [ remaining: Int, set: Set(UNBOND), id: Int ];
+                                         F(record, bond) == 
+                                         IF record.remaining > 0
+                                         THEN 
+                                          IF record.remaining > amount
+                                          THEN [ remaining |-> record.remaining - amount,
+                                                 set |-> record.set \union {[id |-> record.id, amount |-> bond.amount, start |-> bond.epoch, end |-> e]},
+                                                 id |-> record.id + 1 ]
+                                          ELSE [ remaining |-> 0,
+                                                 set |-> record.set \union {[id |-> record.id, amount |-> record.remaining, start |-> bond.epoch, end |-> e]},
+                                                 id |-> record.id + 1 ]    
+                                         ELSE [ remaining |-> 0,
+                                                set |-> record.set,
+                                                id |-> record.id ]
+                                        IN ApaFoldSet(F, [ remaining |-> amount, set |-> {}, id |-> idUnbonds], setBonds)
```

Comment:
I feel like this chunk is not correct.
You initialized `ApaFoldSet` with the value `[ remaining |-> amount, set |-> {}, id |-> idUnbonds]`, therefore, `record.remaining = amount` in the first iteration. In particular, this makes the `IF record.remaining > amount` check immediately false, so this stutters after 1 step with 
```
[ remaining |-> 0,
  set |-> {
    [ id |-> idUnbonds, 
      amount |-> amount, 
      start |-> bond.epoch, 
      end |-> e]
  },
  id |-> idUnbonds + 1 
]    
```
being the final value, for one arbitrarily selected `bond \in setBonds`.
The core issue is the seeded `remaining |-> amount`. It is also probably the case that you don't need to fold over records for this, all you need is the`set` value, because IIUC, `record.remaining`  is always reduced by exactly `amount`, so you can calculate the number of iterations (`min(|setBonds|, ceil(initialAmount / amount)`), which is also the final `id` delta.

Reply:
Good catch! This is a typo. it should be `IF record.remaining > bond.amount`. I've fixed it.